### DISCUSSION TITLE
Limit on palette size for -remap?

### DISCUSSION THREAD

**Discussion by oleglk** [ImageMagick/ImageMagick#7525] [Help] [answered]
### ImageMagick version

7.1.1-3

### Operating system, version and so on

Win 11

### Description

What is the maximum number of colors in supplied palette file for "-remap" action?
==============================
TRY_REMAP>"C:/Program Files/Imagemagick_711_3/convert"  ..\DSC00033.jpg  -dither FloydSteinberg  -remap  c:\Oleg\GitWork\RMA\TMP\all_balanced_colors.tif  -quality 92  .\DSC00033_r.jpg
convert: memory allocation failed `..\DSC00033.jpg' @ error/colormap.c/AcquireImageColormap/123.
convert: memory allocation failed `..\DSC00033.jpg' @ error/quantize.c/SetImageColormap/3980.

TRY_REMAP>"C:/Program Files/Imagemagick_711_3/identify"  ..\DSC00033.jpg
..\DSC00033.jpg JPEG 1620x1080 1620x1080+0+0 8-bit sRGB 354047B 0.000u 0:00.000

TRY_REMAP>"C:/Program Files/Imagemagick_711_3/identify"  -verbose -unique   c:\Oleg\GitWork\RMA\TMP\all_balanced_colors.tif
Image:
  Filename: c:\Oleg\GitWork\RMA\TMP\all_balanced_colors.tif
...
  Geometry: 1x9786628+0+0
  Units: PixelsPerInch
  Colorspace: sRGB
  Type: TrueColorAlpha
  Base type: TrueColor
...
  Colors: 9786628
  Rendering intent: Perceptual
  Gamma: 0.454545
  Chromaticity:
    red primary: (0.64,0.33)
    green primary: (0.3,0.6)
    blue primary: (0.15,0.06)
    white point: (0.3127,0.329)
  Matte color: grey74
  Background color: white
  Border color: srgb(223,223,223)
  Transparent color: black
...
  Filesize: 37.3336MiB
  Number pixels: 9.78663M
...
  Version: ImageMagick 7.1.1-3 Q16 x64 c5d5e71:20230311 https://imagemagick.org

**Comment by snibgo** [comment]
You have v7. `convert` is now deprecated, and you should use `magick` instead.

IM is trying to allocate memory, and failing. The amount of memory it wants is roughly 16 times the number of colours, so about 157MB in your case.

How much memory does your computer have? How much *free* memory? Are you running any other programs at the same time, eg web browsers?

**Reply by oleglk** [reply]
Here is a similar command, although on a smaller :) computer - with 12G RAM.
I looked at the Task Manager while it ran, and the smallest amount of free memory was 4G.
(The palette file is the same of course.)
=================================================
TRY>c:\ANY\Tools\ImageMagick-7.1.1-34\magick.exe  ..\DSC00033.jpg  -remap ..\..\..\TMP\all_balanced_colors.tif  -quality 90  DSC00033__rcb.jpg
magick.exe: unable to create colormap `..\DSC00033.jpg' @ error/colormap.c/AcquireImageColormap/123.
magick.exe: memory allocation failed `..\DSC00033.jpg' @ error/quantize.c/SetImageColormap/4112.

**Comment by snibgo** [comment]
Hmm. There seems to be a problem, possibly a bug. I can reproduce it on my Windows 11 laptop, with 16 GB memory of which 10 GB is free, IM v7.1.1-20:
```
magick rose: -remap hald:7 null:

magick: unable to create colormap `ROSE' @ error/colormap.c/AcquireImageColormap/123.
magick: memory allocation failed `ROSE' @ error/quantize.c/SetImageColormap/4006.
```
Smaller halds, eg `hald:6` work fine. `hald:7` or larger fail.

`hald:7` is only 117649 unique colours, so the requested memory allocation should not be excessive.

What does `magick -list resource` say for you? For me, it says:
```
Resource limits:
  Width: 18.014399PP
  Height: 18.014399PP
  Area: 67.793584GP
  List length: unlimited
  Memory: 15.784424GiB
  Map: 31.568848GiB
  Disk: unlimited
  File: 2400
  Thread: 7
  Throttle: 0
  Time: unlimited
```
... so that isn't causing the problem.

**Reply by oleglk** [reply]
I have a smaller memory limit that doesn't depend on what else is being run:
=====================
\TRY>c:\ANY\Tools\ImageMagick-7.1.1-34\magick.exe  -list resource
Resource limits:
  Width: 72.0576PP
  Height: 72.0576PP
  Area: 25.2237GP
  List length: unlimited
  Memory: 5.87286GiB
  Map: 11.7457GiB
  Disk: unlimited
  File: 1536
  Thread: 3
  Throttle: 0
  Time: unlimited

**Comment by fmw42** [comment]
This fails for me also on IM 7.1.1.36 Mac OSX Ventura

```
magick rose: -remap hald:7 null:
magick: unable to create colormap `ROSE' @ error/colormap.c/AcquireImageColormap/123.
magick: memory allocation failed `ROSE' @ error/quantize.c/SetImageColormap/4112.
```

```
Resource limits:
  Width: 36.0288PP
  Height: 36.0288PP
  Area: 68.7195GP
  List length: unlimited
  Memory: 16GiB
  Map: 32GiB
  Disk: unlimited
  File: 192
  Thread: 1
  Throttle: 0
  Time: unlimited

```

**Comment by snibgo** [comment]
IM has a hard limit of 65536.

The problem is in `colormap.c`, function `AcquireImageColormap()`:
```
  if (colors > MaxColormapSize)
    {
      image->colors=0;
      image->storage_class=DirectClass;
      ThrowBinaryException(ResourceLimitError,"UnableToCreateColormap",
        image->filename);
    }
```

Where `magickcore\magick-type.h` has, for Q16 and above:
```
 #define MaxColormapSize  65536UL
```

I can't see a good reason for imposing a hard limit. Larger colormaps will make some operations slow, but that isn't a good reason for prohibiting them.

I am investigating the possibility of removing this limit.

**Comment by snibgo** [comment]
@oleglk: Please post your remapping image `all_balanced_colors.tif`, and your source image `DSC00033.jpg` although that is less important. You might zip them and upload them to this site.

I want to test my proposed solution with your remapping image.

**Reply by oleglk** [reply]
In case upload to this site fails, you can find these files here:
https://www.dropbox.com/scl/fo/6hdtm3xbiszdj37ufc284/AHgarm9O56paRAQ3ttO740c?rlkey=zbpnz7lmfcxhtjs0pfiq87xo8&st=28ijn8vl&dl=0

And anyway, thanks for your effort !

**Comment by oleglk** [comment]
Here you go:
=========

[all_balanced_colors.zip](https://github.com/user-attachments/files/16481140/all_balanced_colors.zip)

![DSC00033](https://github.com/user-attachments/assets/864cd8a8-836c-4397-9ebf-bc8403b38046)

**Comment by snibgo** [comment]
I have tried removing the restriction on `MaxColormapSize`, but that causes a crash in other processing, and I haven't tracked down where that occurs. So I don't have a solution to this problem. Sorry.

**Comment by urban-warrior** [comment]
Thank you for reporting the issue. We have successfully reproduced it and are actively working on a patch to resolve it. You can expect this patch to be merged into the [main GIT branch](https://github.com/ImageMagick/ImageMagick), later today. As part of our commitment to quality, this fix will also be included in the upcoming [beta releases of ImageMagick](https://imagemagick.org/archive/beta/) by tomorrow. Your patience and feedback are greatly appreciated.

**Reply by oleglk** [reply]
Thanks for the work you did, but... I don't have the environment to build Imagemagick, and investing into it for one-time-use would be an overkill.
So I'd prefer waiting for portable binary release.

**Reply by dlemstra** [reply]
There was a new release this weekend that should include these changes.

**Comment by oleglk** [comment]
OK, I tried it.

The goof news - it finishes and produces a valid image file with most pixels getting expected colors.
The bad news:  some pixels have unlisted colors.

==============================
$ /d/ANY/tmp/IM7/ImageMagick-7.1.1-38/magick.exe  jerry.jpg   -dither FloydSteinberg -remap /c/Oleg/GitWork/RMA/TMP/all_balanced_colors.txt  -quality 92  jerry_r.jpg

$ /d/ANY/tmp/IM7/ImageMagick-7.1.1-38/magick.exe  jerry.jpg    -format "%[pixel:u.p{966,646}]\n" info:
srgb(210,88,73)
$ /d/ANY/tmp/IM7/ImageMagick-7.1.1-38/magick.exe  jerry_r.jpg    -format "%[pixel:u.p{966,646}]\n" info:
srgb(232,108,100)

The rgb(232,108,100) color doesn't appear in the color-list provided:
$ grep "(232,108,100,"  /c/Oleg/GitWork/RMA/TMP/all_balanced_colors.txt
$

**Reply by urban-warrior** [reply]
ImageMagick supports a maximum of 65,535 colors. If this limit is exceeded, ImageMagick reduces the number of unique colors to below 65,535. This process is lossy and may introduce unexpected colors. One alternative is to throw an exception and exit, as done previously. Another option is to ensure the remap image contains fewer than 65,535 unique colors.

**Comment by snibgo** [comment, marked as answer]
You save the remapped result to `jerry_r.jpg`. Writing to JPEG is lossy, which means colour values are changed. Please try again to a lossless format such as PNG.

**Reply by oleglk** [reply]
You are right - that's the reason!
Thanks a lot!

**Comment by oleglk** [comment]
OK, I tried it.

The goof news - it finishes and produces a valid image file with most pixels getting expected colors.
The bad news:  some pixels have unlisted colors.

But let me first clean my side up.
Could you please lay-down the requirements for palette file? For instance, what about ordering of the colors? Then, are repeated appearances allowed?

The proof for unlisted colors:
==============================
$ /d/ANY/tmp/IM7/ImageMagick-7.1.1-38/magick.exe  jerry.jpg   -dither FloydSteinberg -remap /c/Oleg/GitWork/RMA/TMP/all_balanced_colors.txt  -quality 92  jerry_r.jpg

$ /d/ANY/tmp/IM7/ImageMagick-7.1.1-38/magick.exe  jerry.jpg    -format "%[pixel:u.p{966,646}]\n" info:
srgb(210,88,73)
$ /d/ANY/tmp/IM7/ImageMagick-7.1.1-38/magick.exe  jerry_r.jpg    -format "%[pixel:u.p{966,646}]\n" info:
srgb(232,108,100)

The rgb(232,108,100) color doesn't appear in the color-list provided:
$ grep "(232,108,100,"  /c/Oleg/GitWork/RMA/TMP/all_balanced_colors.txt
$
